src/backend_rest.c

```
@@ -0,0 +1,222 @@
+/* backend_rest.c
+   Rémi Attab (remi.attab@gmail.com), 09 Jun 2016
+   FreeBSD-style copyright and disclaimer apply
+*/
+
+#include "optics.h"
+#include "utils/errors.h"
+#include "utils/lock.h"
+#include "utils/buffer.h"
+#include "utils/crest/crest.h"
+
+#include <string.h>
+
+
+// -----------------------------------------------------------------------------
+// metrics
+// -----------------------------------------------------------------------------
+
+struct metric
+{
+    char *key;
+    enum optics_lens_type type;
+    union optics_poll_value value;
+};
+
+struct metrics
+{
+    size_t len;
+    size_t cap;
+    struct metric data[];
+};
+
+enum { metrics_init_cap = 128 };
+
+static void metrics_free(struct metrics *metrics)
+{
+    if (!metrics) return;
+
+    for (size_t i = 0; i < metrics->len; ++i)
+        free(metrics->data[i].key);
+
+    free(metrics);
+}
+
+static struct metrics *metrics_append(struct metrics *metrics, const struct optics_poll *poll)
+{
+    if (!metrics) {
+        metrics = malloc(sizeof(struct metrics) + sizeof(struct metric) * metrics_init_cap);
```

This could be made more similar to the realloc code that follows by setting `->cap` first, and using that instead of `metrics_init_cap`. Also, do you have a philosophy about how to handle failing calls to `malloc` and friends? (The policy could quite well be "let it crash", of course.)